hc-dpki/src/keypair.rs

```diff
@@ -0,0 +1,528 @@
+use crate::{
+    bundle,
+    holochain_sodium::{aead, kx, random::random_secbuf, secbuf::SecBuf, sign},
+    util,
+};
+use holochain_core_types::{agent::KeyBuffer, error::HolochainError};
+use rustc_serialize::json;
+use std::str;
+
+pub struct Keypair {
+    pub pub_keys: String,
+    pub sign_priv: SecBuf,
+    pub enc_priv: SecBuf,
+}
+
+pub const SEEDSIZE: usize = 32 as usize;
+
+impl Keypair {
+    /// derive the pairs from a 32 byte seed buffer
+    ///  
+    /// @param {SecBuf} seed - the seed buffer
+    pub fn new_from_seed(seed: &mut SecBuf) -> Self {
+        let mut seed = seed;
+        let mut sign_public_key = SecBuf::with_insecure(sign::PUBLICKEYBYTES);
+        let mut sign_secret_key = SecBuf::with_secure(sign::SECRETKEYBYTES);
+        let mut enc_public_key = SecBuf::with_insecure(kx::PUBLICKEYBYTES);
+        let mut enc_secret_key = SecBuf::with_secure(kx::SECRETKEYBYTES);
+
+        sign::seed_keypair(&mut sign_public_key, &mut sign_secret_key, &mut seed).unwrap();
+        kx::seed_keypair(&mut seed, &mut enc_public_key, &mut enc_secret_key).unwrap();
+
+        Keypair {
+            pub_keys: util::encode_id(&mut sign_public_key, &mut enc_public_key),
+            sign_priv: sign_secret_key,
+            enc_priv: enc_secret_key,
+        }
+    }
+
+    /// get the keypair identifier string
+    ///
+    /// @return {SecBuf}
+    pub fn get_id(&mut self) -> String {
+        return self.pub_keys.clone();
+    }
+
+    /// generate an encrypted persistence bundle
+    ///
+    /// @param {SecBuf} passphrase - the encryption passphrase
+    ///
+    /// @param {string} hint - additional info / description for the bundle
+    pub fn get_bundle(
+        &mut self,
+        passphrase: &mut SecBuf,
+        hint: String,
+    ) -> Result<bundle::KeyBundle, HolochainError> {
+        let mut passphrase = passphrase;
+        let bundle_type: String = "hcKeypair".to_string();
+        let skk = KeyBuffer::with_corrected(&self.pub_keys)?;
+        let sk = skk.get_sig() as &[u8];
+        let ekk = KeyBuffer::with_corrected(&self.pub_keys)?;
+        let ek = ekk.get_enc() as &[u8];
+        let mut sk_buf = SecBuf::with_insecure(32);
+        let mut ek_buf = SecBuf::with_insecure(32);
+        util::convert_array_to_secbuf(&sk, &mut sk_buf);
+        util::convert_array_to_secbuf(&ek, &mut ek_buf);
+
+        let pw_sign_pub: bundle::ReturnBundleData = util::pw_enc(&mut sk_buf, &mut passphrase);
+        let pw_enc_pub: bundle::ReturnBundleData = util::pw_enc(&mut ek_buf, &mut passphrase);
+
+        let pw_sign_priv: bundle::ReturnBundleData =
+            util::pw_enc(&mut self.sign_priv, &mut passphrase);
+        let pw_enc_priv: bundle::ReturnBundleData =
+            util::pw_enc(&mut self.enc_priv, &mut passphrase);
```

Comment:
It's too inefficient to encrypt these all separately. I think we need to serialize the underlying private data and then encrypt the blob in one go.

since we should know the sizes of things, we can pack them together, something like:

```rust
let bundle = SecBuf::with_secure(
  1 // version byte
  + 32 // sign pubkey
  + 32 // enc pubkey
  + 64?? // sign privkey
  + 64?? // enc privkey
);
{
  let mut bundle = bundle.write_lock();
  bundle[0] = 1; // bundle version number byte
  // write in sk starting at byte 1
  // write in self.sign_priv at proper offset
  // write in ek at proper offset
  // write in self.enc_priv at proper offset
}

let bundle_data = util::pw_enc(&mut bundle, &mut passphrase);
```

Unpacking, would be the reverse, and lots of `array_ref![]`s

Reply:
resolving this, see below: https://github.com/holochain/holochain-rust/pull/863/files#r248417780